src/lib/server/controllers/github.ts

```
@@ -231,4 +240,58 @@ export class GithubController extends HTMLController {
   }
+
+  /*
+   * Returns repo contents of a user-supplied GitHub URL.
+   * Attempts the full path, then progressively less nested paths until valid contents are found.
+   * e.g. the following paths are attempted in order:
+   * `digicatapult/dtdl-visualisation-tool/tree/main/src/lib`
+   * `digicatapult/dtdl-visualisation-tool/tree/main/src`
+   * `digicatapult/dtdl-visualisation-tool/tree/main/.`
+   * `digicatapult/dtdl-visualisation-tool` (returns branches)
+   */
+  @SuccessResponse(200, '')
+  @Get('/navigate')
+  public async navigate(@Query() url: string, @Request() req: express.Request): Promise<HTML | void> {
+    const octokitToken = req.signedCookies[octokitTokenCookie]
+    if (!octokitToken) {
+      this.setStatus(302)
+      this.setHeader('HX-Redirect', authRedirectURL(`/github/picker`))
+      return
+    }
+    const safeUrl = url.replace(/\/$/, '') // remove any trailing slash
+    const match = safeUrl.match(
+      /^(?:https:\/\/(?:www\.)?github\.com\/)?([^/]+)\/([^/]+)(?:\/(?:tree\/)?([^/]+)(?:\/(.+))?)?$/
+    )
+    const [, owner, repo, branch, path] = match || []
+
+    if (!owner || !repo) {
+      throw new GithubReqError(`Invalid URL: ${url}`)
+    }
+
+    if (path && branch) {
+      // Generate all possible paths and attempt each, starting from the most nested
+      const paths = path.split('/').map((_, i, arr) => arr.slice(0, arr.length - i).join('/'))
+      for (const p of paths) {
+        const contents = await this.attemptNavigation(this.contents(owner, repo, p, branch, req))
+        if (contents) return contents
+      }
+    }
+
+    if (branch) {
+      const contents = await this.attemptNavigation(this.contents(owner, repo, '.', branch, req))
+      if (contents) return contents
+    }
+
+    return this.branches(owner, repo, 1, req)
+  }
+
+  async attemptNavigation(nav: Promise<HTML | void>): Promise<HTML | void> {
+    try {
+      return await nav
+    } catch (e) {
+      if (e instanceof GithubReqError) return
```

this is interesting control flow. So if navigation fails we through and therefore the function returns `undefined`. Then we're looping on the possible paths to find the longest one possible? Then if no path is valid we try the branch on it's own and if that fails we return the branch list? 

If that's all correct a comment here might be useful. I'm not sure I can think of a cleaner way of doing this

---

Also do all instances of `GithubReqError` mean we should loop like this? Is there a possibility that something else has gone wrong?

---

a `GithubReqError` could happen for some other reason than the `path` given by the user is incorrect. Token could expire, GitHub could be down. If that happened the code reaches the final line `return this.branches(owner, repo, 1, req)` which would throw the same previously ignored error -> user gets a toast saying `GitHub API request failed`

---

I could make a more specific error for when the GitHub request 404s and only ignore that one

---

added